### PR TITLE
OBSDOCS-290 - Loki Log Access

### PR DESCRIPTION
**Pull request by libander**
Version(s): 4.12+

Issue: https://issues.redhat.com/browse/OBSDOCS-290

Link to docs preview: https://66726--docspreview.netlify.app/openshift-enterprise/latest/logging/cluster-logging-loki#logging-loki-log-access_cluster-logging-loki

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: logging/cluster-logging-loki.adoc
Comment (abrennan89) [inline logging/cluster-logging-loki.adoc:40]: This isn't showing up in the preview because it should be
```
include::modules/logging-loki-log-access.adoc[leveloffset=+1]
```

File: modules/logging-loki-log-access.adoc
Comment (abrennan89) [inline modules/logging-loki-log-access.adoc:7]: If this has a title like "Configuring" and shows how to configure settings it should probably be a procedure with numbered steps instead of a `CONCEPT` module.

File: modules/logging-loki-log-access.adoc
Comment (kabirbhartiRH) [inline modules/logging-loki-log-access.adoc:48]: We could also have a RoleBinding to 'cluster-logging-application-view' ClusterRole which is namespace restricted. This can be used for providing logs access to regular users who should not have Cluster-wide access.

File: modules/logging-loki-log-access.adoc
Comment (kabirbhartiRH) [inline modules/logging-loki-log-access.adoc:50]: 3) Specifies all authenticated users/groups

**Comment by kabirbhartiRH** [comment]
@periklis Should we also add a note that the old RBAC's will not be present on a fresh installation of Logging 5.8 but it would be when Logging is upgraded from 5.7 to newer versions?

File: modules/logging-loki-log-access.adoc
Comment (libander) [reply modules/logging-loki-log-access.adoc:7]: How does this differ from `logging-loki-restart-hardening.adoc` ?

File: logging/cluster-logging-loki.adoc
Comment (libander) [reply logging/cluster-logging-loki.adoc:40]: Whoops, thanks!

File: modules/logging-loki-log-access.adoc
Comment (libander) [reply modules/logging-loki-log-access.adoc:50]: @kabirbhartiRH - Would `Specifies all authenticated users regardless of group membership` work? Groups aren't something I think are thought of as authenticating the same way as a user.

**Comment by libander** [comment]
> @periklis Should we also add a note that the old RBAC's will not be present on a fresh installation of Logging 5.8 but it would be when Logging is upgraded from 5.7 to newer versions?

This is a great question. Do we have upgrade path sorted out yet?

**Review by JoaoBraveCoding — COMMENTED** [review]
So I feel like we might need to expand a bit on some of the details, here's some things I believe should be covered:
- For users upgrading from 5.y to 5.8, log access permissions are the same that were used in previous logging versions. This means that every authenticated user has permission to read logs in their namespaces, this is supported by the clusterrole `logging-application-logs-reader` and clusterrolebinding `logging-all-authenticated-application-logs-reader`. If users want to start using the new permission schema they should delete these resources.
- With 5.8 we introduce 3 new ClusterRoles that grant access to different types of logs. 1. `cluster-logging-application-view` grants permissions to read application logs. 2. `cluster-logging-infrastructure-view` grants permissions to read infrastructure logs. 3. `cluster-logging-audit-view` grants permission to read audit logs.
- In 5.8 the logging operator assumes a least privilege principle where access to logs is not granted to regular users out of the box. Instead, OpenShift admins are required to grant access to users as they see fit. For instance, a similar permission schema to older logging versions could be achieved by creating a clusterrolebinding of `cluster-logging-application-view` to the `system:authenticated` Group.
- In 5.8 we consider Admin every user who belongs to a group that is listed in the `adminGroups` section of LokiStack. Admin users have access to all types of logs in all namespaces.

cc @periklis if I missed something or phrased it wrong

File: modules/logging-loki-log-access.adoc
Comment (xperimental) [reply modules/logging-loki-log-access.adoc:50]: I think this might be a misunderstanding between Kabir and you. You correctly stated that `system:authenticated` is a group containing `all authenticated users` but Kabir (probably) wanted to hint at the possibility that you can list both User and Group resources as subjects.

Is it maybe a good idea to list both a user and a group in the example?

File: modules/logging-loki-log-access.adoc
Comment (xperimental) [inline modules/logging-loki-log-access.adoc:28]:
```suggestion
<2> Overrides the default groups (`system:cluster-admins`, `cluster-admin`, `dedicated-admin`)
```

---

Does it make sense to also mention that it's possible to disable the admin groups by providing an empty list:

```
adminGroups: []
```

**Comment by xperimental** [comment]
@JoaoBraveCoding made some good points, just have one minor correction on the last one (emphasis mine):

> * In 5.8 we consider Admin every user who belongs to a group that is listed in the `adminGroups` section of LokiStack. **Admin users have access to all types of logs in all namespaces.**

Admin users still have to pass the SubjectAccessReview to be allowed to read from any of the tenants (`application`, `audit` and `infrastructure`), so the ClusterRoleBinding to the ClusterRole is still needed (creating a RoleBinding for a user that's in the adminGroups does not make any sense). Being part of the `adminGroups` just disables the "namespace check", so that if you have cluster-wide access to the application logs, no further checks are done and you can just read all the application logs. (the "namespace check" is currently only done for application logs, the `adminGroups` setting thus has no effect for `infrastructure` or `audit` logs.

File: modules/logging-loki-log-access.adoc
Comment (abrennan89) [reply modules/logging-loki-log-access.adoc:7]: That doc just shows an example of how to configure a setting in one file, this one shows settings for LokiStack CR plus different role bindings, so it seems like those things need to be configured together to make this work? Maybe not, though.

Comment (libander) [reply modules/logging-loki-log-access.adoc:7]: There is no action the user needs to take outlined here. The examples are exposures of default values requiring no configuration by the user. This is a 'hey this is a new setting, here's where it is & what the fields mean' module not a 'how to do specific use-case xyz'.

File: modules/logging-loki-log-access.adoc
Comment (libander) [reply modules/logging-loki-log-access.adoc:50]: Thanks for the clarification :) On second pass, I think removing callout 3 works, as the 2nd callout covers it, given that the `kind:` is `group`. Thoughts?

Comment (kabirbhartiRH) [reply modules/logging-loki-log-access.adoc]: Yes. looks okay to me

File: modules/logging-loki-log-access.adoc
Comment (kabirbhartiRH) [inline modules/logging-loki-log-access.adoc:56]: Does it make sense to set the name here as a custom admin group 'custom-admin-group' on here as mentioned on the above LokiStack CR.

**Comment by libander** [comment]
Note: check if this module should be added to the upgrading section prior to merging.

File: modules/logging-loki-log-access.adoc
Comment (xperimental) [inline modules/logging-loki-log-access.adoc:19]: "the all application logs in all namespaces" only applies if the user also has the required Kubernetes permission for the application tenant. For some of these groups this always applies (for example `system:cluster-admins` and `cluster-admin` usually have _all privileges_ by design).

```suggestion
Users who are members of any group specified in the `adminGroups` field of the LokiStack CR are considered admins. Admin users have access to all application logs in all namespaces, if they also get assigned the `cluster-logging-application-view` role.
```

File: modules/logging-loki-log-access.adoc
Comment (xperimental) [inline modules/logging-loki-log-access.adoc:64]: Is it possible to have some visible distinction between "configure access cluster-wide and configure access per namespace"? Currently this file reads like "first do adminGroups, then do cluster access, then namespaced". I don't think this gets the message across that these are two methods independent from each other and that the `adminGroups` customization is "probably" not needed for most users.

File: modules/logging-loki-log-access.adoc
Comment (JoaoBraveCoding) [inline modules/logging-loki-log-access.adoc:54]:
```suggestion
  name: cluster-logging-application-view # <1>
```

**Comment by libander** [comment]
@anpingli / @QiaolingTang - LGTM please.

**Review by periklis — COMMENTED** [review]
LGTM

**Comment by michaelryanpeter** [comment]
@libander As a reminder, please wait until your PRs have been through peer review and verified by QE before adding them to the merge review queue.

**Comment by libander** [comment]
> @libander As a reminder, please wait until your PRs have been through peer review and verified by QE before adding them to the merge review queue.

Sorry, I was trying to force it to remove the do not merge flag that seems to be stuck on from when it was in draft status.

**Comment by anpingli** [comment]
LGTM

**Review by mjpytlak — COMMENTED** [review]
@libander while the content itself lgtm, I have a few suggestions on how the module is constructed and might benefit from a few more new modules that are based on what you have already written.

**Comment by libander** [comment]
> @libander while the content itself lgtm, I have a few suggestions on how the module is constructed and might benefit from a few more new modules that are based on what you have already written.

I see what you mean, and considered it, but these aren't procedures. This just covers the different approaches cluster admins can take to fine tune access to logs.  RBAC has its own section of the documentation regarding the creation of ClusterRoleBindings and RoleBindings, if anything this could be perhaps a concept module and a reference module for each of the 3 sections, but that seems a bit granular.

**Comment by mjpytlak** [comment]
Closing/reopening to give Travis a nudge.

**Comment by mjpytlak** [comment]
Closing/reopening to give Travis a nudge.

File: modules/logging-loki-log-access.adoc
Comment (jeana-redhat) [inline modules/logging-loki-log-access.adoc:74]: Noticed a typo, not sure if it's "you have large deployments" or "you have a large deployment" though

**Review by jeana-redhat — COMMENTED** [review]
Noticed a small typo (sorry!) while looking for merge. Otherwise LGTM

/remove-label merge-review-in-progress
/remove-label merge-review-needed

**Comment by libander** [comment]
> Noticed a small typo (sorry!) while looking for merge. Otherwise LGTM
That's what merge reviews are for! Fixed, thank you!  :)

**Review by jeana-redhat — APPROVED** [review]
LGTM

/remove-label merge-review-in-progress
/remove-label merge-review-needed

**Comment by jeana-redhat** [comment]
C'mon, Travis

**Comment by jeana-redhat** [comment]
Build plz

**Pull request merged**
